text/0030-protocol-messages.md

```diff
@@ -0,0 +1,248 @@
+    /// The size of the buffer should be within the range returned by the `size_range` method.
+    fn from_bytes(bytes: &[u8]) -> Self;
+
+    /// Returns the size of the message.
+    fn size(&self) -> usize;
+
+    /// Serializes a message into a byte buffer.
+    /// Panics if the provided buffer has an invalid size.
+    /// The size of the buffer should be equal to the one returned by the `size` method.
+    fn into_bytes(self, bytes: &mut [u8]);
+}
+```
+
+**Notes**:
+- `size_range` returns an allowed range for the message size because some parts of some messages can be trimmed. It is
+  used to check if a message coming from a transport layer has a valid size. More details on compression below;
+- `from_bytes`/`into_bytes` panic if incorrectly used, only the following safe TLV module should directly use them;
+- `into_bytes` does not allocate a buffer because the following TLV protocol implies concatenating a header inducing
+  another allocation. Since this is a hot path, a slice of an already allocated buffer for both the header and payload
+  is expected; hence, limiting the amount of allocation to the bare minimum;
+
+## Type-length-value protocol
+
+The [type-length-value](https://en.wikipedia.org/wiki/Type-length-value) module is a safe layer on top of the messages.
+It allows serialization/deserialization to/from a byte buffer ready to be sent/received to/from a transport layer by
+prepending or reading a header containing the type and length of the payload.
+
+### Header
+
+```rust
+/// A header for the type-length-value encoding.
+struct Header {
+    /// Type of the message.
+    message_type: u8,
+    /// Length of the message.
+    message_length: u16,
+}
+```
+
+### Methods
+
+```rust
+/// Deserializes a TLV header and a byte buffer into a message.
+/// * The advertised message type should match the required message type.
+/// * The advertised message length should match the buffer length.
+/// * The buffer length should be within the allowed size range of the required message type.
+fn tlv_from_bytes<M: Message>(header: &Header, bytes: &[u8]) -> Result<M, TlvError> {
+    ...
+}
+
+/// Serializes a TLV header and a message into a byte buffer.
+fn tlv_into_bytes<M: Message>(message: M) -> Vec<u8> {
+    ...
+}
+```
+
+## Messages
+
+Since the various types of messages are constructed with different kind of data, there can not be a single constructor
+signature in the `Message` trait. Implementations are then expected to provide a convenient `new` method to build them.
+
+### Endianness
+
+All multi-byte number fields of the messages of the protocol are represented as [big-endian](https://en.wikipedia.org/wiki/Endianness).
+
+### Derived traits
+
+The following traits are expected to be derived by every `Message` implementation:
+
+- `Default` which is very convenient for the implementation of `Message::from_bytes`;
+- `Clone` which is necessary to provide ownership in the context of a message broadcast;
```

Comment:
Is broadcasting something that's expected to be supported by all message types (and can trivially be supported by most message types)? If so, a `trait Message: Clone` constraint might be beneficial such that generic code doesn't require extra trait constraints to do broadcasting.

Reply:
If this is the case, I wonder whether having `into_bytes` take `&self` is a good idea such that broadcasting can be done without cloning the contents (which might be a slow operation if the message constraints heap-allocated data structures).

Reply:
It's actually not expected to be supported by all messages. It just made it simpler to have a generic broadcast function instead of repeating it for the messages that actually do.

Reply:
I like the idea of having it as a trait constraint.

Reply:
Is it bad to put it as a constraint even if not used by *all* impls ? It just adds a bit of work on compile time but goes away, right ?

Reply:
Turns out I can just clone the bytes and not the entire message.